### PR TITLE
feat: added reusable component

### PR DESCRIPTION
**Pull request by raaymax**
I'm not entirely satisfied with how the selection of this component works, but on the other hand, it operates similarly to the repeater component. Additionally, I'm unsure if there's any sense in using dynamic options for the parent component ID in this case. It might be easier to just manually copy and paste the component IDs.

### CODE REVIEW

**Comment by ramedina86** [comment]
That's a good point about ids... maybe just a "copy" icon next to the id for convenience? What do you think?

**Comment by raaymax** [comment]
https://github.com/streamsync-cloud/streamsync/assets/649972/733611ea-59d8-4810-814c-1477bb08fc8d

File: ui/src/builder/BuilderCopyText.vue [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+.tooltip .tooltiptext::after {
+	content: "";
+	position: absolute;
+	top: 100%;
+	left: 50%;
+	margin-left: -5px;
+	border-width: 5px;
+	border-style: solid;
+	border-color: #555 transparent transparent transparent;
```
Comment (ramedina86) [inline ui/src/builder/BuilderCopyText.vue:55]: Can we avoid hardcoding here? i.e. `var(--builderActionOngoingColor);`

I'm also seeing there's a very minor difference in colour between the two.

File: ui/src/builder/BuilderCopyText.vue [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	display: inline-block;
+}
+
+.tooltip .tooltiptext {
+	visibility: hidden;
+	background-color: var(--builderActionOngoingColor);
+	color: #fff;
+	text-align: center;
+	border-radius: 20px;
+	padding: 5px 10px;
```
Comment (ramedina86) [inline ui/src/builder/BuilderCopyText.vue:36]: I try quite hard to use multiples of `4px` in margins, paddings, border-radiuses, widths, etc e.g. `4px 12px`. Reason is maintaining a grid and hopefully one day migrating to a more sophisticated way of managing spacing. It's less of an issue in Builder, but still I'd prefer it.

File: ui/src/builder/BuilderCopyText.vue [hunk 3]
```diff
@@ -0,0 +1,69 @@
+<template>
+	<span class="tooltip">
```
Comment (ramedina86) [inline ui/src/builder/BuilderCopyText.vue:2]: If we can center it on long ids that'd be good, but not a deal breaker at all.

Screenshot 2024-03-12 at 16 11 23

File: ui/src/builder/BuilderCopyText.vue [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+<style scoped>
+.tooltip {
+	position: relative;
+	display: inline-block;
+}
+
+.tooltip .tooltiptext {
+	visibility: hidden;
+	background-color: var(--builderActionOngoingColor);
+	color: #fff;
```
Comment (ramedina86) [inline ui/src/builder/BuilderCopyText.vue:33]: Please use `--builderBackgroundColor`, looks better and if one day we get dark mode or do a Barbie edition we're set.

File: ui/src/core_components/other/CoreReusable.vue [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+function render() {
+	if (!fields.parentId.value) {
+		vnode.value = h("div", "No component selected");
+		return;
+	}
+	const reusedNode = renderProxiedComponent(fields.parentId.value, 0);
+	vnode.value = reusedNode;
+}
+
+watch([fields.parentId], render);
```
Comment (ramedina86) [inline ui/src/core_components/other/CoreReusable.vue:44]: Better to use `{ immediate: true }` when setting up the watcher.

File: ui/src/core_components/other/CoreReusable.vue [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+};
+</script>
+<script setup lang="ts">
+const fields = inject(injectionKeys.evaluatedFields);
+const renderProxiedComponent = inject(injectionKeys.renderProxiedComponent);
+const vnode = ref(h("div", "No component selected"));
+
+function render() {
+	if (!fields.parentId.value) {
+		vnode.value = h("div", "No component selected");
```
Comment (ramedina86) [inline ui/src/core_components/other/CoreReusable.vue:37]: Can we make this behave like Empty _Message_ for example? Also please specify for what purpose i.e. "No component specified for reuse.".

It shouldn't show on Preview/run (use `isBeingEdited`) injectable.

File: ui/src/core_components/other/CoreReusable.vue [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		return;
+	}
+	const reusedNode = renderProxiedComponent(fields.parentId.value, 0);
```
Comment (ramedina86) [inline ui/src/core_components/other/CoreReusable.vue:40]: Please check if the component exists and if not show "The id specified for reuse doesn't match any component." or similar. Also shouldn't show on Preview/run.

File: ui/src/core_components/other/CoreReusable.vue [hunk 8]
```diff
@@ -0,0 +1,54 @@
+<template>
+	<div class="CoreReuse">
+		<component :is="vnode" />
```
Comment (ramedina86) [inline ui/src/core_components/other/CoreReusable.vue:3]: 🚨🚨🚨 I think the whole thing will blow up when using _Sidebar_, which is one of the main use cases.

**Review by ramedina86 — COMMENTED** [review]
I tried this via Scalingo. Experience is great already.

Please address the comment on _Sidebar_ first, it might be a pain.

I've added some comments to improve it a bit further, plus my usual nitpicking when it comes to any ui work.

File: ui/src/core_components/other/CoreReusable.vue (hunk 8 above)
Comment (raaymax) [reply ui/src/core_components/other/CoreReusable.vue:3]: Yes, it exploded for sidebar.

**Comment by FabienArcellier** [comment]
Great work @raaymax . I love this component. It works well and the setup experience is interesting. I would like to suggest an improvement for the edit mode. I think the selection experience would be improved if the user couldn't select inside the component. Currently it's jumping around the component tree chaotically.

Peek 2024-03-13 08-54

Would it be possible to add an invisible overlay above the component and once the component is selected, it become visible as grey overlay and display a button that moves on click the developper to the original component?

File: ui/src/builder/BuilderCopyText.vue (hunk 3 above)
Comment (raaymax) [reply ui/src/builder/BuilderCopyText.vue:2]: It is centered to span containing this text, only thing I can do here is adding:
```
white-space: nowrap;
```
but not sure if this is the way?

**Comment by raaymax** [comment]
@FabienArcellier I don't think selecting nested components this way would be possible.
I could try to block selection inside of reusable component entirely and add button "go to component" or something to settings panel on right. What do you think about it?

**Comment by FabienArcellier** [comment]
> @FabienArcellier I don't think selecting nested components this way would be possible. I could try to block selection inside of reusable component entirely and add button "go to component" or something to settings panel on right. What do you think about it?

That would be awesome. That's what I thought I asked

**Pull request merged**